File: 22-generate-parentheses/22-generate-parentheses.cpp
class Solution {
public:
    
    vector <string> result;
    
    void generate(string s, int left, int right)
    {
        if(left == 0 && right == 0)
            result.push_back(s);
        if(left > 0)
            generate(s + "(", left - 1, right);
        if(right > left)
            generate(s + ")", left, right - 1);
        
    }
    vector<string> generateParenthesis(int n) 
    {
        generate("", n, n);
        return result;
    }
};

File: 22-generate-parentheses/22-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        output = []
        def backtrack(s, l, r):
            if l == 0 and r == 0:
                output.append(s)
            if l != 0:
                backtrack(s + "(", l - 1, r)
            if r > l:
                backtrack(s + ")", l, r - 1)
        backtrack("", n, n)
        return output
        